File: transferUV/src/transferUV.cpp
// transferUV.cpp
//
// Copyright (c) 2018 Michitaka Inoue
//
// This software is released under the MIT License.
// http://opensource.org/licenses/mit-license.php


#include "transferUV.h"
#include <iostream>
#include <maya/MArgDatabase.h>
#include <maya/MArgList.h>
#include <maya/MGlobal.h>
#include <maya/MSelectionList.h>
#include <maya/MString.h>
#include <maya/MSyntax.h>

// Flags for this command
static const char* sourceUvSetFlag = "-suv";
static const char* sourceUvSetFlagLong = "-sourceUvSet";
static const char* targetUvSetFlag = "-tuv";
static const char* targetUvSetFlagLong = "-targetUvSet";
static const char* sourceMeshFlag = "-sm";
static const char* sourceMeshFlagLong = "-sourceMesh";
static const char* targetMeshFlag = "-tm";
static const char* targetMeshFlagLong = "-targetMesh";

TransferUV::TransferUV()
{
}

TransferUV::~TransferUV()
{
}

MSyntax TransferUV::newSyntax()
{
    MSyntax syntax;

    MStatus status;
    status = syntax.addFlag(sourceUvSetFlag, sourceUvSetFlagLong, MSyntax::kString);
    CHECK_MSTATUS_AND_RETURN(status, syntax);

    status = syntax.addFlag(targetUvSetFlag, targetUvSetFlagLong, MSyntax::kString);
    CHECK_MSTATUS_AND_RETURN(status, syntax);

    status = syntax.addFlag(sourceMeshFlag, sourceMeshFlagLong, MSyntax::kString);
    CHECK_MSTATUS_AND_RETURN(status, syntax);

    status = syntax.addFlag(targetMeshFlag, targetMeshFlagLong, MSyntax::kString);
    CHECK_MSTATUS_AND_RETURN(status, syntax);

    return syntax;
}

MStatus TransferUV::doIt(const MArgList& args)
{
    MStatus status;

    // if (args.length() != 1)
    // {
    //     MGlobal::displayError("Need 1 arg!");
    //     return MStatus::kFailure;
    // }

    MArgDatabase argData(syntax(), args);

    // arg
    // MString argument = args.asString(0, &status);
    // if (status != MS::kSuccess) {
    //     return MStatus::kFailure;
    // }
    // CHECK_MSTATUS_AND_RETURN_IT(status);

    if (argData.isFlagSet(sourceUvSetFlag)) {
        status = argData.getFlagArgument(sourceUvSetFlag, 0, sourceUvSet);
        CHECK_MSTATUS_AND_RETURN_IT(status);
    }

    if (argData.isFlagSet(targetUvSetFlag)) {
        status = argData.getFlagArgument(targetUvSetFlag, 0, targetUvSet);
        CHECK_MSTATUS_AND_RETURN_IT(status);
    }

    if (argData.isFlagSet(sourceMeshFlag)) {
        status = argData.getFlagArgument(sourceMeshFlag, 0, sourceMesh);
    } else {
        MGlobal::displayError(
            "Source mesh is not specified. -sm flag is required.");
        return MS::kFailure;
    }

    if (argData.isFlagSet(targetMeshFlag)) {
        status = argData.getFlagArgument(targetMeshFlag, 0, targetMesh);
    } else {
        MGlobal::displayError(
            "Target mesh is not specified. -tm flag is required.");
        return MS::kFailure;
    }

    MString info = "Copying uv from " + sourceUvSet + " to " + targetUvSet;
    MGlobal::displayInfo(info);

    MSelectionList mList;
    mList.add(sourceMesh);
    mList.add(targetMesh);

    mList.getDagPath(0, sourceDagPath);
    mList.getDagPath(1, targetDagPath);

    return redoIt();
}

MStatus TransferUV::redoIt()
{
    MStatus status;

    MFnMesh sourceFnMesh(sourceDagPath);
    MFnMesh targetFnMesh(targetDagPath);
    
    // Get current UV sets and keep them to switch back to them at the last
    MString currentSourceUVSet = sourceFnMesh.currentUVSetName();
    MString currentTargetUVSet = targetFnMesh.currentUVSetName();

    // Switch to uv sets specified in flags before starting transfer process
    // This is required because MFnMesh does not work properly with
    // "Non-current" UV sets
    sourceFnMesh.setCurrentUVSetName(sourceUvSet);
    targetFnMesh.setCurrentUVSetName(targetUvSet);
    
    MString* sourceUvSetPtr = &sourceUvSet;
    MString* targetUvSetPtr = &targetUvSet;

    MFloatArray sourceUarray;
    MFloatArray sourceVarray;
    MIntArray sourceUvCounts;
    MIntArray sourceUvIds;

    // Get mesh information for each
    status = sourceFnMesh.getUVs(sourceUarray, sourceVarray, sourceUvSetPtr);
    if (status != MS::kSuccess) {
        MGlobal::displayError("Failed to get source UVs");
        CHECK_MSTATUS_AND_RETURN_IT(status);
    }
    status = targetFnMesh.getUVs(originalUarray, originalVarray, targetUvSetPtr);
    if (status != MS::kSuccess) {
        MGlobal::displayError("Failed to get original UVs");
        CHECK_MSTATUS_AND_RETURN_IT(status);
    }
    status = sourceFnMesh.getAssignedUVs(sourceUvCounts, sourceUvIds, sourceUvSetPtr);
    if (status != MS::kSuccess) {
        MGlobal::displayError("Failed to get source assigned UVs");
        CHECK_MSTATUS_AND_RETURN_IT(status);
    }
    status = targetFnMesh.getAssignedUVs(originalUvCounts, originalUvIds, targetUvSetPtr);
    if (status != MS::kSuccess) {
        MGlobal::displayError("Failed to get original assigned UVs");
        CHECK_MSTATUS_AND_RETURN_IT(status);
    }

    // Resize source uv array so it becomes same size as number of targets uv
    unsigned int targetNumUVs = targetFnMesh.numUVs();
    if (targetNumUVs > sourceUarray.length()) {
        sourceUarray.setLength(targetNumUVs);
        sourceVarray.setLength(targetNumUVs);
    }

    status = targetFnMesh.setUVs(sourceUarray, sourceVarray, targetUvSetPtr);
    if (MS::kSuccess != status) {
        MGlobal::displayError("Failed to set source UVs");
        return status;
    }

    status = targetFnMesh.assignUVs(sourceUvCounts, sourceUvIds, targetUvSetPtr);
    if (MS::kSuccess != status) {
        MGlobal::displayError("Failed to assign source UVs");
        return status;
    }

    // Switch back to originals
    sourceFnMesh.setCurrentUVSetName(currentSourceUVSet);
    targetFnMesh.setCurrentUVSetName(currentTargetUVSet);
    
    return MS::kSuccess;
}

MStatus TransferUV::undoIt()
{
    MStatus status;

    MFnMesh undoMesh(targetDagPath);
    MGlobal::displayInfo(targetDagPath.fullPathName());

    MString* targetUvSetPtr = &targetUvSet;

    // Resize original uv array so it becomes same size as current num uvs
    unsigned int numUVs = undoMesh.numUVs();
    if (numUVs > originalUarray.length()) {
        originalUarray.setLength(undoMesh.numUVs());
        originalVarray.setLength(undoMesh.numUVs());
    }

    status = undoMesh.setUVs(originalUarray, originalVarray, targetUvSetPtr);
    if (MS::kSuccess != status) {
        MGlobal::displayError("Failed to set original UVs");
        return status;
    } else {
        MGlobal::displayError("OK");
    }

    status = undoMesh.assignUVs(originalUvCounts, originalUvIds, targetUvSetPtr);
    if (MS::kSuccess != status) {
        MGlobal::displayError("Failed to assgin original UVs");
        return status;
    }

    return MS::kSuccess;
}

bool TransferUV::isUndoable() const
{
    return true;
}

void* TransferUV::creator()
{
    return new TransferUV;
}

File: transferUV/py/transferUV.py
""" Yet another UV transfer command """

import sys
from maya.api import OpenMaya


kPluginCmdName = "transferUV"
kPluginVersion = "0.0.2"
kPluginAuthor = "Michi Inoue"
kSourceUvSetFlag = "-suv"
kSourceUvSetFlagLong = "-sourceUvSet"
kTargetUvSetFlag = "-tuv"
kTargetUvSetFlagLong = "-targetUvSet"
kSourceMeshFlag = "-sm"
kSourceMeshFlagLong = "-sourceMesh"
kTargetMeshFlag = "-tm"
kTargetMeshFlagLong = "-targetMesh"


class TransferUV(OpenMaya.MPxCommand):

    def __init__(self):
        super(TransferUV, self).__init__()

        self.sourceMesh = None
        self.targetMesh = None
        self.sourceUvSet = None
        self.targetUvSet = None

    def doIt(self, args):

        # Parse the arguments.
        argData = OpenMaya.MArgDatabase(self.syntax(), args)
        try:
            self.cmdArg = argData.commandArgumentString(0)
        except RuntimeError:
            pass

        if argData.isFlagSet(kSourceMeshFlag):
            argList = argData.getFlagArgumentList(kSourceMeshFlag, 0)
            self.sourceMesh = argList.asString(0)

        if argData.isFlagSet(kTargetMeshFlag):
            argList = argData.getFlagArgumentList(kTargetMeshFlag, 0)
            self.targetMesh = argList.asString(0)

        if argData.isFlagSet(kSourceUvSetFlag):
            argList = argData.getFlagArgumentList(kSourceUvSetFlag, 0)
            self.sourceUvSet = argList.asString(0)

        if argData.isFlagSet(kTargetUvSetFlag):
            argList = argData.getFlagArgumentList(kTargetUvSetFlag, 0)
            self.targetUvSet = argList.asString(0)

        if self.sourceUvSet is None:
            self.sourceUvSet = "map1"

        if self.targetUvSet is None:
            self.targetUvSet = "map1"

        mSel = OpenMaya.MGlobal.getActiveSelectionList()
        numSel = mSel.length()

        if self.sourceMesh is None and numSel < 2:
            OpenMaya.MGlobal.displayWarning(
                (""" Source mesh and target mesh are not specifiied, or not enough objs are """
                 """selected. Specify source and target in the arguments or select two objs"""))
            return

        if self.sourceMesh is None:
            try:
                self.sourceDagPath = mSel.getDagPath(0)
                print("Source mesh is not given in the arguments. Using {} as a source mesh.".format(self.sourceDagPath.fullPathName()))
            except RuntimeError:
                OpenMaya.MGlobal.displayWarning(
                    "Failed to set source DagPath. Select source mesh first, then target mesh")
                return
        else:
            try:
                mSel.add(self.sourceMesh)
                self.sourceDagPath = mSel.getDagPath(0)
            except RuntimeError:
                OpenMaya.MGlobal.displayWarning("Object doesn't exist")
                return

        if self.targetMesh is None:
            try:
                self.targetDagPath = mSel.getDagPath(1)
                print("Target mesh is not given in the arguments. Using {} as a target mesh.".format(self.targetDagPath.fullPathName()))
            except RuntimeError:
                OpenMaya.MGlobal.displayWarning(
                    "Failed to set target DagPath. Select source mesh first, then target mesh")
                return
        else:
            try:
                mSel.add(self.targetMesh)
                self.targetDagPath = mSel.getDagPath(1)
            except RuntimeError:
                OpenMaya.MGlobal.displayWarning("Object doesn't exist")
                return

        print("Copying uv from {} to {}".format(
            self.sourceDagPath.fullPathName(),
            self.targetDagPath.fullPathName()))

        self.redoIt()

    def redoIt(self):

        sourceFnMesh = OpenMaya.MFnMesh(self.sourceDagPath)
        targetFnMesh = OpenMaya.MFnMesh(self.targetDagPath)

        sourceUvCounts = OpenMaya.MIntArray()
        sourceUvIds = OpenMaya.MIntArray()

        sourceUarray, sourceVarray = sourceFnMesh.getUVs(self.sourceUvSet)
        self.originalUarray, self.originalVarray = targetFnMesh.getUVs(self.targetUvSet)

        sourceUvCounts, sourceUvIds = sourceFnMesh.getAssignedUVs()
        self.originalUvCounts, self.originalUvIds = targetFnMesh.getAssignedUVs()

        # Clear target UVs before transfer in order to shrink the uvs array if number of source UVs
        # are smaller than that of target UVs
        targetFnMesh.clearUVs(self.targetUvSet)

        targetFnMesh.setUVs(sourceUarray, sourceVarray, self.targetUvSet)
        targetFnMesh.assignUVs(sourceUvCounts, sourceUvIds, self.targetUvSet)

    def undoIt(self):
        undoMesh = OpenMaya.MFnMesh(self.targetDagPath)

        # Clear target UVs before transfer in order to shrink the uvs array if number of source UVs
        # are smaller than that of target UVs
        undoMesh.clearUVs(self.targetUvSet)
        undoMesh.setUVs(self.originalUarray, self.originalVarray, self.targetUvSet)
        undoMesh.assignUVs(self.originalUvCounts, self.originalUvIds, self.targetUvSet)

    def isUndoable(self):
        return True

    @staticmethod
    def cmdCreator():
        return TransferUV()


def syntaxCreator():
    """ Syntax creator

    Return:
        syntax (OpenMaya.MSyntax): return value

    """
    syntax = OpenMaya.MSyntax()
    syntax.addArg(OpenMaya.MSyntax.kString)
    syntax.addFlag(kSourceUvSetFlag, kSourceUvSetFlagLong, OpenMaya.MSyntax.kString)
    syntax.addFlag(kTargetUvSetFlag, kTargetUvSetFlagLong, OpenMaya.MSyntax.kString)
    syntax.addFlag(kSourceMeshFlag, kSourceMeshFlagLong, OpenMaya.MSyntax.kString)
    syntax.addFlag(kTargetMeshFlag, kTargetMeshFlagLong, OpenMaya.MSyntax.kString)
    return syntax


def maya_useNewAPI():
    """
    The presence of this function tells Maya that the plugin produces, and
    expects to be passed, objects created using the Maya Python API 2.0.
    """
    pass


def initializePlugin(mObject):
    """ Initialize the script plug-in

    Args:
        mObject (OpenMaya.MObject):

    """
    mplugin = OpenMaya.MFnPlugin(mObject, kPluginAuthor, kPluginVersion, "Any")
    try:
        mplugin.registerCommand(
            kPluginCmdName, TransferUV.cmdCreator, syntaxCreator)
    except Exception:
        sys.stderr.write("Failed to register command: %s\n" % kPluginCmdName)
        raise


def uninitializePlugin(mObject):
    """ Uninitialize the script plug-in

    Args:
        mObject (OpenMaya.MObject):

    """
    mplugin = OpenMaya.MFnPlugin(mObject)
    try:
        mplugin.deregisterCommand(kPluginCmdName)
    except Exception:
        sys.stderr.write("Failed to unregister command: %s\n" % kPluginCmdName)
        raise
